File: Travelling Salesman Problem 3_python/src/AntColonyOptimization.py
import os
import time
from multiprocessing import Pool
from multiprocessing import cpu_count

# import game
from Ant import Ant
from Maze import Maze
from PathSpecification import PathSpecification


class AntColonyOptimization:
    """
    Class representing the first assignment. Finds shortest path between two points in a maze according to a specific path specification.
    """

    def __init__(self, maze, ants_per_gen, generations, q, evaporation, minconstant):
        """
        Constructs a new optimization object using ants.
         :param: maze the maze .
         :param: antsPerGen the amount of ants per generation.
         :param: generations the amount of generations.
         :param: Q normalization factor for the amount of dropped pheromone
         :param: evaporation the evaporation factor.
        """
        self.maze = maze
        self.ants_per_gen = ants_per_gen
        self.generations = generations
        self.q = q
        self.evaporation = evaporation
        self.minconstant = minconstant

    def get_route(self, pathspec):
        # print(f"Ant {i+1} of generation {g+1}")
        ant = Ant(self.maze, pathspec)
        return ant.find_route()

    def get_route_wrapper(self, args):
        return self.get_route(*args)

    def find_shortest_route(self, path_specification):
        """
        Loop that starts the shortest path process
        :param: spec Spefication of the route we wish to optimize
        :return: ACO optimized route
        """
        largelist = []
        p = Pool(cpu_count())

        for g in range(self.generations):
            print(f"Starting gen {g+1}")
            start_time = time.time()

            routelist = p.map(self.get_route_wrapper, zip([path_specification] * self.ants_per_gen))
            diff = time.time() - start_time

            print(f"Finished generation {g+1} in {diff:.2f} seconds with an avg of {sum(route.size() for route in routelist)/len(routelist)}.\nEvaporating previous pheromone")

            self.maze.evaporate(self.evaporation, self.minconstant)
            print(f"Adding pheromone")
            start_time = time.time()
            self.maze.add_pheromone_routes(routelist, self.q)
            diff = time.time() - start_time
            print(f"Finished pheromones in {diff:.2f} seconds")
            largelist.append(routelist)
            # print("\n".join(str(el) for el in self.maze.pheromones))

        shortestroute = largelist[0][0]
        for row in largelist:
            for route in row[1:]:
                # print(str(route))
                if route.shorter_than(shortestroute):
                    shortestroute = route

        self.maze.reset()
        return shortestroute


# Driver function for Assignment 1
if __name__ == "__main__":
    # parameters
    antsingen = 50  # ants per generation
    no_gen = 50  # amount of generations
    q = 1600  # constant to multiply the amount of pheromone
    evap = 0.5  # 1 - evaporation constant
    factor = 0.3
    gamify = False
    mazetype = "medium"
    # convergence criterion:
    # e.g. no imporvement in n steps

    # construct the optimization objects
    print(os.path.abspath(f"./../data/{mazetype} maze.txt"))
    maze = Maze.create_maze(f"./../data/{mazetype} maze.txt")
    spec = PathSpecification.read_coordinates(f"./../data/{mazetype} coordinates.txt")
    aco = AntColonyOptimization(maze, antsingen, no_gen, q, evap, factor)
    print(maze.walls)
    # if gamify:
    #     # startup game
    #     # app = game.App(maze, spec.start)
    #     app.on_init()
    #     app.on_render()
    # else:
    #     app = None
    # app.on_execute()

    # save starting time
    start_time_all = int(round(time.time() * 1000))

    # run optimization
    # shortest_route = aco.find_shortest_route(spec, app)
    shortest_route = aco.find_shortest_route(spec)
    # if gamify:
    #     app.on_cleanup()
    # print time taken
    print("Time taken: " + str((int(round(time.time() * 1000)) - start_time_all) / 1000.0))

    # save solution
    shortest_route.write_to_file(f"./../data/{mazetype}_solution.txt")

    # print route size
    print("Route size: " + str(shortest_route.size()))

File: Travelling Salesman Problem 3_java/src/main/java/AntColonyOptimization.java
import aliceinnets.python.jyplot.JyPlot;

import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.logging.Level;
import java.util.logging.Logger;


/**
 * Class representing the first assignment. Finds shortest path between two points in a maze according to a specific
 * path specification.
 */
public class AntColonyOptimization {
    public static final int ELITISM = 10;
    private static Level logLevel = Level.ALL;
    private List<Integer> averages;
    private List<Integer> smallest;
    private Logger logger = Logger.getLogger(AntColonyOptimization.class.getName());
    private int antsPerGen;
    private int convergenceRuns;
    private int generations;
    private double Q;
    private double factor;
    private double evaporation;
    private ExecutorService executor;
    private Maze maze;

    /**
     * Constructs a new optimization object using ants.
     *
     * @param maze        the maze .
     * @param antsPerGen  the amount of ants per generation.
     * @param generations the amount of generations.
     * @param Q           normalization factor for the amount of dropped pheromone
     * @param evaporation the evaporation factor.
     */
    public AntColonyOptimization(Maze maze, int antsPerGen, int generations, double Q, double evaporation, int convergenceRuns) {
        this.maze = maze;
        this.antsPerGen = antsPerGen;
        this.generations = generations;
        this.convergenceRuns = convergenceRuns;
        this.averages = new ArrayList<>(generations);
        this.smallest = new ArrayList<>(generations);
        this.factor = 0.00001;
        this.executor = Executors.newFixedThreadPool(antsPerGen);
        this.Q = Q;
        this.evaporation = evaporation;
        this.logger.setLevel(logLevel);
    }

    /**
     * Driver function for Assignment 1
     */
    public static void main(String[] args) throws FileNotFoundException {
        //parameters
        int ants = 100;
        int noGen = 150;
        int convergenceRuns = 150;

        double Q = 1600;
        double evap = 0.5;
        String mazetype = "hard";

        //construct the optimization objects
        Maze maze = Maze.createMaze(String.format("./data/%s maze.txt", mazetype));
        PathSpecification spec = PathSpecification.readCoordinates(String.format("./data/%s coordinates.txt", mazetype));
        AntColonyOptimization aco = new AntColonyOptimization(maze, ants, noGen, Q, evap, convergenceRuns);


        //save starting time
        long startTime = System.currentTimeMillis();

        //run optimization
        Route shortestRoute = aco.findShortestRoute(spec);

        //print time taken
        System.out.println("Time taken: " + ((System.currentTimeMillis() - startTime) / 1000.0) + " seconds");
        JyPlot plt = new JyPlot();
        for (int i = 0; i < aco.smallest.size(); i++) {
            plt.scatter(i, aco.smallest.get(i), 5, "#ff0000");
            plt.scatter(i, aco.averages.get(i), 5, "#00ff00");
        }
        plt.title(String.format("Steps over generations with %d ants at %d", ants, shortestRoute.size()));

        plt.ylabel("Number of steps");
        plt.xlabel("Generation");
//        plt.legend();

        plt.show();
        plt.exec();
        //save solution
        shortestRoute.writeToFile(String.format("./data/%s_solution.txt", mazetype));

        //print route size
        System.out.println("Route size: " + shortestRoute.size());
        aco.close();
    }

    public void close() {
        this.executor.shutdownNow();
    }

    /**
     * Loop that starts the shortest path process
     *
     * @param spec Spefication of the route we wish to optimize
     * @return ACO optimized route
     */
    public Route findShortestRoute(PathSpecification spec) {
        List<Route> routes = new ArrayList<>();
        Route shortestRoute = null;
        int shortRouteCount = 0;
        int averageShortest = 0;
        for (int g = 0; g < this.generations; g++) {
//            logger.log(Level.INFO, "Starting gen " + (g));
            long start = System.currentTimeMillis();

            try {
                List<Callable<Route>> croutes = new ArrayList<>(this.antsPerGen);
                for (int i = 0; i < this.antsPerGen; i++) {
                    croutes.add(() -> new Ant(maze, spec).findRoute());
                }
                List<Future<Route>> froutes = executor.invokeAll(croutes);
                for (Future<Route> fr : froutes) {
                    routes.add(fr.get());
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
            long diff = System.currentTimeMillis() - start;
            try {
                routes.sort(Route::compareTo);
            } catch (NullPointerException ignored) {
            }

            if (shortestRoute == null || routes.get(0).shorterThan(shortestRoute)) {
                shortestRoute = routes.get(0);
                shortRouteCount = 0;

            } else if (shortRouteCount++ > this.convergenceRuns) {
                System.out.printf("Converged on %d after %d runs%n", shortestRoute.size(), g);
                break;
            }

            int sum = routes.stream().mapToInt(Route::size).sum();
            int average = sum / this.antsPerGen;
            this.averages.add(average);
            this.smallest.add(routes.get(0).size());
            System.out.printf("Finished gen %d in %d millis with average of %d (best %d)%n", g, diff, sum / this.antsPerGen, shortestRoute.size());
            if (average == shortestRoute.size() && ++averageShortest > this.convergenceRuns) {
                System.out.printf("Converged on %d after %d runs%n", shortestRoute.size(), g);
            } else if (average != shortestRoute.size()) {
                averageShortest = 0;
            }


            this.maze.evaporate(this.evaporation, this.factor);
            weightedPheromones(routes, this.Q);
            this.maze.addPheromoneRoute(shortestRoute, this.Q);
            routes.clear();
        }
        maze.printPheromones(shortestRoute, spec);
        maze.reset();
        return shortestRoute;
    }

    private void weightedPheromones(List<Route> routes, double startq) {
        int i = 1;
        for (Route r : routes.subList(0, ELITISM)) {
            this.maze.addPheromoneRoute(r, (startq * i++) / ELITISM);
        }
    }

}
